default/python/AI/ColonisationAI.py

```diff
@@ -1310,3 +1224,265 @@ def get_first_column_value(x):
+            try:
+                best_ship = next(design for design in fo.getEmpire().availableShipDesigns
+                                 if "SD_OUTPOST_BASE" == fo.getShipDesign(design).name)
+                debug("Falling back to base design SD_OUTPOST_BASE")
+            except StopIteration:
+                # fallback design not available
+                return False
+
+        # enqueue the design at the source planet
+        retval = fo.issueEnqueueShipProductionOrder(best_ship, self.source)
+        print "Enqueueing Outpost Base at %s for %s with result %s" % (
+            universe.getPlanet(self.source), universe.getPlanet(self.target), retval)
+
+        if not retval:
+            warn("Failed to enqueue outpost base at %s" % universe.getPlanet(self.source))
+            return False
+
+        self.base_enqueued = True
+        return True
+
+    @property
+    def base_assigned(self):
+        if self.fid == INVALID_ID:
+            return False
+
+        fleet = fo.getUniverse().getFleet(self.fid)
+        if fleet:
+            return True
+
+        debug("The fleet assigned to the OrbitalColonizationPlan doesn't exist anymore.")
+        self.fid = INVALID_ID
+        return False
+
+    @property
+    def score(self):
+        if self.__last_score_update != fo.currentTurn():
+            self.__update_score()
+        return self.__score
+
+    def __update_score(self):
+        planet_score = evaluate_planet(self.target, MissionType.OUTPOST, None)
+        for species in empire_colonizers:
+            this_score = evaluate_planet(self.target, MissionType.COLONISATION, species)
+            planet_score = max(planet_score, this_score)
+        self.__last_score_update = fo.currentTurn()
+        self.__score = planet_score
+
+    def is_valid(self):
+        universe = fo.getUniverse()
+
+        # make sure target is valid
+        target = universe.getPlanet(self.target)
+        if target is None or (not target.unowned) or target.speciesName:
+            return False
+
+        # make sure source is valid
+        source = universe.getPlanet(self.source)
+        if not (source and source.ownedBy(fo.empireID()) and source.speciesName in empire_colonizers):
+            return False
+
+        # appears to be valid
+        return True
+
+
+class OrbitalColonizationManager(object):
+    """
+    The OrbitalColonizationManager handles orbital colonization for the AI.
+
+    :type _colonization_plans: dict[int, OrbitalColonizationPlan]
+    :type num_enqueued_bases: int
+    """
+    def __init__(self):
+        self._colonization_plans = {}
+        self.num_enqueued_bases = 0
+
+    def get_targets(self):
+        """
+        Return all planets for which an orbital colonization plan exists.
+
+        :rtype: list[int]
+        """
+        return self._colonization_plans.keys()
+
+    def delete_plan(self, target_id):
```

Comment:
Unused method should be removed.   Please check usage of the other new methods.

Reply:
> Unused method should be removed. 

While lack of current use is certainly reason to think twice about whether a certain piece of code might be essentially unneeded distracting complexity, I don't see lack of current use as an automatically sufficient reason to always delete the code, not in our environment at least.  Morlic is presenting a toolset here which by itself accomplishes certain things, and he apparently contemplates the likelihood of one of us or someone else further developing more complex approaches that could include deleting an orbital construction plan.  That capability seems like a good one to be able to support.  Making a rule that we couldn't ever provide any support for more complex planning than what is currently implemented seems guaranteed to slow down development.

Reply:
This code is never executed, so it is not tested.   
Writing unused code is waste of time. 
Reviewing unused code is waste of time.
Supporting unused code is waste of time.

When you will need new public method, you will add public method that you need.  Using method that was written 10 years ago, and did not executed since may give you a surprise.

This method is simple code and it have small chance to be broken.   But it is easy to implement it when needed.  For complex code chance that it will require changes when it will be finally used is about 100%. This number is based on my personal experience as full day developer in big project.

In terms of master, we don't remove any code, we just don't add things that not needed now.

Reply:
The function was in use in a previous implementation stage but I removed it for now. The reimplementation would be trivial.